py/dml/ctree.py

```diff
@@ -4787,30 +4790,41 @@ def mkCast(site, expr, new_type):
             raise ETEMPLATEUPCAST(site, "object", new_type)
         else:
             return mkTraitUpcast(site, expr, real.trait)
+
+    if (compat.dml12_misc in dml.globals.enabled_compat
+        and isinstance(expr, InterfaceMethodRef)):
+        # Workaround for SIMICS-9868
+        return mkLit(site, "%s->%s" % (
+            expr.node_expr.read(), expr.method_name), new_type)
+
+    if isinstance(expr, NonValue):
+        raise expr.exc()
     old_type = safe_realtype(expr.ctype())
     if (dml.globals.compat_dml12_int(site)
         and (isinstance(old_type, (TStruct, TVector))
              or isinstance(real, (TStruct, TVector)))):
         # these casts are permitted by C only if old and new are
         # the same type, which is useless
         return Cast(site, expr, new_type)
-    if isinstance(real, TStruct):
-        if isinstance(old_type, TStruct) and old_type.label == real.label:
-            return expr
-        raise ECAST(site, expr, new_type)
-    if isinstance(real, TExternStruct):
-        if isinstance(old_type, TExternStruct) and old_type.id == real.id:
-            return expr
+    if isinstance(real, (TVoid, TArray, TFunction)):
         raise ECAST(site, expr, new_type)
-    if isinstance(real, (TVoid, TArray, TVector, TTraitList, TFunction)):
+    if old_type.cmp(real) == 0:
```

Comment:
I just realized this path is disgustingly not always correct due to bitfields. A 64-bit bitfields is completely type compatible with `uint64`, even in the sense as laid out by the docstring of `DMLType.cmp`, and yet DMLC will treat them differently w.r.t. `mkSubRef`, so just spitting out the same expression `mkRValue`-d is not correct. I suppose we could have an ad-hoc exception for those cases.

Reply:
Alternatively, instead of doing `mkRValue`, add an arg for `Cast` that tells it not to spit out a cast in generated C. Or an alternate class to `Cast` that does the same thing. Or use `Cast` anyway.

Reply:
This also really highlights the difference between type _equality_ and type _compatibility_. We've gotten away with treating them as one and the same, but in this instance it falls apart. Does that mean we should introduce a separate notion for type _equality..._?

Reply:
I think we'll be forced to, if we're not willing to treat bitfields as type-compatible with the corresponding `uintX` type. Consider idempotent externs; an `extern` with a bitfields arg can use `{}` initialized syntax, but an `extern` with a `uintX` arg cannot. We cannot consider two declarations differing in that aspect equal to one another, even though their signatures are completely compatible. There's a similar issues with method declarations and overrides.

Reply:
The notion of type equality would thus be type compatibility with an additional restriction: that DMLC treats the type identically in all ways that are visible to the modeler.

Reply:
Added an ad-hoc exception for the moment.

Reply:
I don't think compatibility between `uintN` and `bitfields N` is important -- one can surely come up with use cases, but I can't see how they would be common (e.g., did you ever get a uint64 pointer that you want to pass to a function that takes a bitfields pointer?). If making them incompatible removes the need for a "type equality" concept, then it's most likely worth it. Especially since we now have a compat feature to hijack.

Reply:
If it's easy to introduce uintN vs bitfeldsN incompatibility separately, without the full type revamp, then it would be nice to run that with jenkins, to grasp how much this part would break.

Reply:
hah my ad-hoc exception wasn't good enough to clear jenkins:
```
        vid1 = (cast(&vid1, vlan_tag_t *))->vid;
```
`&vid` is `uint16 *` which is considered compatible with `vlan_tag_t *` and thus `mkRValue` is used, oops.

Anyhoo, I'll try out adding the rule that a bitfields is only considered compatible to another bitfields if they share the same exact members (maybe with some memoization...?) in addition to the other properties like signedness and endianness.